drivers/net/w5500/w5500.c

```diff
@@ -0,0 +1,1099 @@
+int w5500_reset(struct w5500_dev *dev)
+{
+	int ret;
+	uint8_t mr;
+	uint8_t ver;
+
+	if (dev->gpio_reset) {
+		ret = no_os_gpio_set_value(dev->gpio_reset, 0);
+		if (ret)
+			return ret;
+
+		/* Hold reset low for at least 500us per datasheet */
+		no_os_udelay(500);
+
+		ret = no_os_gpio_set_value(dev->gpio_reset, 1);
+		if (ret)
+			return ret;
+	} else {
+		mr = W5500_MR_RST;
+		ret = w5500_reg_write(dev, W5500_COMMON_REG, W5500_MR, &mr, 1);
+		if (ret)
+			return ret;
+	}
+
+	/* Wait for reset to complete */
+	no_os_mdelay(1);
+
+	ret = w5500_reg_read(dev, W5500_COMMON_REG, W5500_MR, &mr, 1);
+	if (ret)
+		return ret;
+
+	if (mr & W5500_MR_RST)
+		return -ETIMEDOUT;
+
+	ret = w5500_reg_read(dev, W5500_COMMON_REG, W5500_VERSIONR, &ver, 1);
+	if (ret)
+		return ret;
+
+	if (ver != W5500_CHIP_VERSION)
+		return -EINVAL;
+
+	return 0;
+}
+
+/***************************************************************************//**
+ * @brief Setup the W5500 chip with basic configuration
+ *
+ * @param dev - The device descriptor
+ *
+ * @return 0 in case of success, negative error code otherwise
+*******************************************************************************/
+int w5500_setup(struct w5500_dev *dev)
+{
+	int ret;
+
+	ret = w5500_reset(dev);
+	if (ret)
+		return ret;
+
+	ret = w5500_set_mac(dev, dev->mac_addr);
+	if (ret)
+		return ret;
+
+	ret = w5500_set_ip(dev, dev->ip_addr);
+	if (ret)
+		return ret;
+
+	ret = w5500_set_subnet(dev, dev->subnet_mask);
+	if (ret)
+		return ret;
+
+	ret = w5500_set_gateway(dev, dev->gateway);
+	if (ret)
+		return ret;
+
+	ret = w5500_tcp_config(dev, dev->retry_time, dev->retry_count);
+	if (ret)
+		return ret;
+
+	return w5500_check_link_status(dev);
+}
+
+/***************************************************************************//**
+ * @brief Initialize the device
+ *
+ * @param device     - Pointer to the device descriptor to be initialized
+ * @param init_param - The device's initialization parameters
+ *
+ * @return 0 in case of success, negative error code otherwise
+*******************************************************************************/
+int w5500_init(struct w5500_dev **device,
+	       struct w5500_init_param *init_param)
+{
+	struct w5500_dev *dev;
+	int ret;
+
+	dev = (struct w5500_dev *)malloc(sizeof(*dev));
+	if (!dev)
+		return -ENOMEM;
+
+	ret = no_os_gpio_get_optional(&dev->gpio_reset, &init_param->gpio_reset);
+	if (ret)
+		goto free_dev;
+
+	ret = no_os_gpio_get_optional(&dev->gpio_int, &init_param->gpio_int);
+	if (ret)
+		goto free_gpio_reset;
+
+	ret = no_os_spi_init(&dev->spi, &init_param->spi_init);
+	if (ret)
+		goto free_gpio_int;
+
+	memcpy(dev->mac_addr, init_param->mac_addr, 6);
+	memcpy(dev->ip_addr, init_param->ip_addr, 4);
+	memcpy(dev->subnet_mask, init_param->subnet_mask, 4);
+	memcpy(dev->gateway, init_param->gateway, 4);
+
+	memcpy(&dev->retry_time, &init_param->retry_time, 1);
+	memcpy(&dev->retry_count, &init_param->retry_count, 1);
```

Comment:
The W5500 seems to have some default values for these that look reasonable. I think these 2 fields in the init parameter should be "optional" in the sense that if they are 0, you don't actually write the register. This way you don't force the users to check the datasheet.